com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java

```diff
@@ -0,0 +1,110 @@
+                    String.format("RestartFarme: cannot find the stack frame with frameID %s", restartFrameArgs.frameId));
+        }
+
+        if (canRestartFrame(context, stackFrameReference)) {
+            try {
+                ThreadReference reference = stackFrameReference.getThread();
+                popStackFrames(context, reference, stackFrameReference.getDepth());
+                stepin(context, reference);
+            } catch (DebugException de) {
+                context.getProtocolServer().sendEvent(new Events.UserNotificationEvent(NotificationType.ERROR, de.getMessage()));
+            }
+            return CompletableFuture.completedFuture(response);
+        } else {
+            context.getProtocolServer().sendEvent(new Events.UserNotificationEvent(NotificationType.ERROR, "Current stack frame doesn't support restart."));
+            context.getProtocolServer().sendEvent(new Events.StoppedEvent("restartframe", stackFrameReference.getThread().uniqueID()));
+            return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.RESTARTFRAME_FAILURE, "Failed to restart the selected stack frame.");
+        }
+    }
+
+    private boolean canRestartFrame(IDebugAdapterContext context, StackFrameReference frameReference) {
+        ThreadReference reference = frameReference.getThread();
+        StackFrame[] frames = context.getStackFrameManager().reloadStackFrames(reference);
+        // Current stack frame cannot be the top frame of the thread:
+        if (frames.length <= frameReference.getDepth() + 1) {
+            return false;
+        }
+
+        // Cannot restart frame involved with native call stacks:
+        for (int i = 0; i <= frameReference.getDepth() + 1; i++) {
+            if (StackFrameUtility.isNative(frames[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void popStackFrames(IDebugAdapterContext context, ThreadReference thread, int depth) throws DebugException {
+        while (depth >= 0) {
+            StackFrame[] frames = context.getStackFrameManager().reloadStackFrames(thread);
+            StackFrameUtility.pop(frames[0]);
+            --depth;
+        }
+    }
+
+    private void stepin(IDebugAdapterContext context, ThreadReference thread) {
+        StepRequest request = DebugUtility.createStepIntoRequest(thread, context.getStepFilters().classNameFilters);
+        context.getDebugSession().getEventHub().stepEvents().filter(debugEvent -> request.equals(debugEvent.event.request())).take(1).subscribe(debugEvent -> {
+            debugEvent.shouldResume = false;
+            // Have to send to events to keep the UI sync with the step in operations:
+            context.getProtocolServer().sendEvent(new Events.StoppedEvent("step", thread.uniqueID()));
+            context.getProtocolServer().sendEvent(new Events.ContinuedEvent(thread.uniqueID()));
```

Comment:
why to send continued event to vscode

Reply:
After the stack operation (pop/stepin), the UI is out of sync with the underlying JVM debugger status. Thus, send continue event to update the VSCode UI to let the VSCode populate the threading information. The continue event will stop at the first line of the stepin method.